processors/contrib/processor.py

```diff
@@ -0,0 +1,139 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import io
+import os
+import re
+import uuid
+import boto3
+import shutil
+import logging
+import zipfile
+import hashlib
+import tempfile
+import requests
+logger = logging.getLogger(__name__)
+
+
+# Module API
+
+def process(conf, conn):
+
+    # Prepare s3 resource
+    resource = boto3.resource('s3',
+        region_name=conf['AWS_S3_REGION'],
+        aws_access_key_id=conf['AWS_ACCESS_KEY_ID'],
+        aws_secret_access_key=conf['AWS_SECRET_ACCESS_KEY'])
+
+    # Iterate contributions mapping
+    for contrib_id, mappings in conf['CONTRIB'].items():
+
+        # Prepare temp directory
+        dirpath = tempfile.mkdtemp()
+
+        # Get contribution and download documents
+        contrib = conn['explorerdb']['data_contributions'].find_one(id=contrib_id)
+        _download_documents(contrib['data_url'], dirpath)
+
+        # Process downloaded documents
+        for filename in os.listdir(dirpath):
+
+            # Extract primary_id
+            (type, primary_id) = _extract_metadata(filename, mappings)
+            if not primary_id:
+                logger.warning('Document "%s" is not in contrib mappings', filename)
+                continue
+
+            # Retrieve trial
+            trial = conn['database']['trials'].find_one(primary_id=primary_id)
+            if not trial:
+                logger.warning('Document "%s" has no matched trial', filename)
+                continue
+
+            # Build document url
+            filepath = os.path.join(dirpath, filename)
+            checksum = _caclulate_checksum(filepath)
+            extension = os.path.splitext(filename)[1]
+            bucket = conf['AWS_S3_BUCKET']
+            key = 'documents/%s%s' % (checksum, extension)
+            url = _generate_url(resource, bucket, key)
+
+            # Save document (if not exists)
+            table = conn['database']['documents']
+            if not table.find_one(trial_id=trial['id'], url=url):
+
+                # Upload document to S3
+                resource.Bucket(bucket).upload_file(filepath, key)
+
+                # Write document to database
+                table.insert({
+                    'id': uuid.uuid4().hex,
```

Comment:
It won't make a difference here, but UUIDv4 wreaks havoc over indexes because their values, by design, are spread over all possible values. Because of this, it's better to use UUIDv1.